File: scripts/build_artwork.py
"""Build the profile's original SVG artwork. Python standard library only."""
from pathlib import Path
from html import escape
from math import sin, cos, pi

ROOT = Path(__file__).resolve().parents[1]
OUT = ROOT / 'assets' / 'profile'
INK, PAPER, MUTED, LINE, LIME = '#101210', '#f1f2e9', '#9ba394', '#30372d', '#d1ff73'

CSS = '''
.orbit{animation:orbit 32s linear infinite;transform-box:fill-box;transform-origin:center}
.float{animation:float 7s ease-in-out infinite}
.flow{stroke-dasharray:12 180;animation:flow 10s linear infinite}
.breathe{animation:breathe 5s ease-in-out infinite}
@keyframes orbit{to{transform:rotate(360deg)}}
@keyframes float{50%{transform:translateY(-9px)}}
@keyframes flow{to{stroke-dashoffset:-384}}
@keyframes breathe{50%{opacity:.45}}
@media(prefers-reduced-motion:reduce){*{animation:none!important}}
'''


def text(x, y, value, size=20, color=PAPER, weight=400, extra=''):
    return f'<text x="{x}" y="{y}" fill="{color}" font-family="Arial,Helvetica,sans-serif" font-size="{size}" font-weight="{weight}" {extra}>{escape(value)}</text>'


def label(x, y, value, color=MUTED, size=13):
    return text(x, y, value, size, color, 400, 'letter-spacing="2"')


def line(x1, y1, x2, y2, color=LINE, extra=''):
    return f'<path d="M{x1} {y1}L{x2} {y2}" stroke="{color}" fill="none" {extra}/>'


def circle(x, y, r, color=LIME, extra=''):
    return f'<circle cx="{x}" cy="{y}" r="{r}" fill="{color}" {extra}/>'


def svg(name, width, height, title, body, description=''):
    value = f'''<svg xmlns="http://www.w3.org/2000/svg" width="{width}" height="{height}" viewBox="0 0 {width} {height}" role="img" aria-labelledby="title desc">
<title id="title">{escape(title)}</title><desc id="desc">{escape(description or title)}</desc>
<defs>
<radialGradient id="halo"><stop stop-color="#384529" stop-opacity=".7"/><stop offset="1" stop-color="{INK}" stop-opacity="0"/></radialGradient>
<linearGradient id="metal" x1="0" y1="0" x2="1" y2="1"><stop stop-color="{PAPER}"/><stop offset=".5" stop-color="#85966b"/><stop offset="1" stop-color="{LIME}"/></linearGradient>
<pattern id="grid" width="40" height="40" patternUnits="userSpaceOnUse"><path d="M40 0H0V40" fill="none" stroke="#232a20" stroke-width=".65"/></pattern>
<clipPath id="bounds"><rect width="{width}" height="{height}" rx="20"/></clipPath>
</defs><style>{CSS}</style>
<g clip-path="url(#bounds)"><rect width="{width}" height="{height}" fill="{INK}"/>{body}</g>
<rect x=".5" y=".5" width="{width-1}" height="{height-1}" rx="20" fill="none" stroke="{LINE}"/>
</svg>'''
    (OUT / name).write_text(value + '\n')


def torus(cx, cy, scale=1):
    """A projected wireframe torus: exact geometry, animated as vector artwork."""
    pieces = [f'<g transform="translate({cx} {cy}) scale({scale})">', '<circle r="260" fill="url(#halo)"/>']
    pieces.append('<g class="float">')
    paths = []
    def point(u, v):
        x, y, z = (133 + 48*cos(v))*cos(u), (133 + 48*cos(v))*sin(u), 48*sin(v)
        # Tilt the ring toward the viewer, then rotate the composition.
        yy = y*.51-z*.86
        return x*.90-yy*.43, x*.43+yy*.90
    for v in range(20):
        points = [point(i*2*pi/100, v*2*pi/20) for i in range(101)]
        d = 'M' + 'L'.join(f'{x:.1f} {y:.1f}' for x,y in points) + 'Z'
        paths.append(f'<path d="{d}" fill="none" stroke="url(#metal)" stroke-opacity=".55" stroke-width=".85"/>')
    for u in range(44):
        points = [point(u*2*pi/44, i*2*pi/32) for i in range(33)]
        d = 'M' + 'L'.join(f'{x:.1f} {y:.1f}' for x,y in points) + 'Z'
        paths.append(f'<path d="{d}" fill="none" stroke="url(#metal)" stroke-opacity=".46" stroke-width=".8"/>')
    pieces.extend(paths)
    pieces.append('</g><g class="orbit">')
    pieces.append('<circle r="203" fill="none" stroke="#697656" stroke-width=".8" stroke-dasharray="2 12"/>')
    pieces.append('<circle r="220" fill="none" stroke="#d1ff73" stroke-opacity=".6" stroke-dasharray="160 1222"/>')
    pieces.append(circle(203,0,4)+circle(-203,0,3,PAPER))
    pieces.append('</g>')
    pieces.append(text(0, 13, 'ZM', 39, PAPER, 700, 'text-anchor="middle" letter-spacing="-3"'))
    pieces.append('</g>')
    return ''.join(pieces)


def hero(mobile=False):
    w,h = (600,780) if mobile else (1200,570)
    b = f'<rect x="{0 if mobile else 700}" width="{w}" height="{h}" fill="url(#grid)"/>'
    b += text(36, 49, 'ZM', 23, LIME, 800, 'letter-spacing="-2"')
    b += label(89,47,'ZAIDMOEN / GITHUB',size=12)
    b += line(36,72,w-36,72)
    if mobile:
        b += label(36,119,'SOFTWARE DEVELOPER',LIME,14)
        b += text(30,209,'ZAID',106,PAPER,800,'letter-spacing="-7"')
        b += text(32,282,'MAYYALLEH',69,PAPER,800,'letter-spacing="-4"')
        b += text(36,332,'Thoughtful interfaces. Solid foundations.',21,MUTED)
        b += torus(300,525,.86)
        b += line(36,730,564,730)
        b += label(36,760,'PALESTINE',size=12)+label(347,760,'BUILD / LEARN / REFINE',size=10)
    else:
        b += label(830,47,'CODE. CRAFT. CURIOSITY.',size=12)
        b += label(42,137,'SOFTWARE DEVELOPER',LIME,14)
        b += text(34,260,'ZAID',144,PAPER,800,'letter-spacing="-9"')
        b += text(36,354,'MAYYALLEH',89,PAPER,800,'letter-spacing="-5"')
        b += text(42,412,'Thoughtful interfaces. Solid foundations.',25,MUTED)
        b += torus(930,289,.93)
        b += line(42,505,1158,505)
        b += label(42,542,'BASED IN PALESTINE',size=12)
        b += label(440,542,'FULL-STACK / BACKEND / APPLIED AI',size=12)
        b += label(976,542,'ALWAYS BUILDING',LIME,12)
    svg('hero-mobile.svg' if mobile else 'hero.svg',w,h,'Zaid Mayyalleh — Software developer in Palestine',b,
        'Thoughtful interfaces. Solid foundations. Full-stack development, backend systems and applied AI. A gently moving wireframe orbit surrounds the ZM monogram. Motion respects reduced-motion preferences.')


def cube(x,y,s=30):
    return f'<g transform="translate({x} {y})"><path d="M0 {-s}L{s} {-s/2} 0 0 {-s} {-s/2}Z" fill="#d1ff73"/><path d="M{-s} {-s/2}L0 0V{s}L{-s} {s/2}Z" fill="#667c40"/><path d="M0 0L{s} {-s/2}V{s/2}L0 {s}Z" fill="#a4cb65"/></g>'


def illustration(kind):
    # All illustrations occupy a 330 x 180 viewport.
    b = ''
    if kind == 'voxel':
        for i in range(6):
            b += line(55+i*22,95+i*11,165+i*22,40+i*11,'#323d29')
            b += line(55+i*22,95-i*11,165+i*22,150-i*11,'#323d29')
        b += cube(162,119,27)+cube(216,119,27)+cube(189,105,27)
        b += '<g class="float">'+cube(162,64,27)+'</g>'
        b += '<path d="M243 43h20V23M263 43l-31-31M82 128H62v20" fill="none" stroke="#d1ff73" stroke-width="2"/>'
    elif kind == 'api':
        for x,y,label_ in [(25,58,'API'),(131,58,'AUTH'),(237,58,'SQL')]:
            b += f'<rect x="{x}" y="{y}" width="70" height="64" rx="10" fill="#1c2318" stroke="#617344"/>'
            b += text(x+35,y+37,label_,15,LIME,600,'text-anchor="middle"')
        b += '<path d="M95 90H131M201 90H237M60 124v25H271v-25" fill="none" stroke="#54643f"/>'
        b += '<path class="flow" d="M60 149H272V90H60" fill="none" stroke="#d1ff73" stroke-width="2"/>'
        b += label(101,32,'REQUEST / RESPONSE',size=10)
    elif kind == 'web':
        b += '<g class="float"><rect x="48" y="20" width="240" height="148" rx="10" fill="#1b2118" stroke="#667a4c"/>'
        b += line(48,44,288,44)+circle(62,32,2)+circle(72,32,2,MUTED)+circle(82,32,2,MUTED)
        b += f'<rect x="203" y="58" width="65" height="7" rx="3" fill="{LIME}"/>'
        for yy,ww in [(76,94),(89,126)]:
            b += f'<rect x="{268-ww}" y="{yy}" width="{ww}" height="4" rx="2" fill="#6d7c5e"/>'
        for xx in [67,137,207]:
            b += f'<rect x="{xx}" y="108" width="61" height="42" rx="4" fill="#2c3921"/>'
        b += '</g><path class="flow" d="M48 179H288" stroke="#d1ff73" stroke-width="2"/>'
    elif kind == 'ml':
        for x,ys in [(65,[50,90,130]),(165,[28,70,112,154]),(265,[66,116])]:
            nx,nys=(165,[28,70,112,154]) if x==65 else (265,[66,116])
            if x<265:
                for y in ys:
                    for ny in nys:
                        b+=line(x,y,nx,ny,'#3b4b2d')
        for x,ys in [(65,[50,90,130]),(165,[28,70,112,154]),(265,[66,116])]:
            for y in ys:
                b+=circle(x,y,7,INK,'stroke="#d1ff73" stroke-width="1.5"')
        b+='<path class="flow" d="M65 90L165 70 265 116" stroke="#d1ff73" fill="none" stroke-width="2"/>'
    return b


PROJECTS = [
    ('api','01','Books Management API','BACKEND ENGINEERING','FastAPI / MySQL / JWT','Relational data. Explicit permissions. Raw SQL.'),
    ('voxel','02','Pinch Voxel Studio','COMPUTER VISION','Python / OpenCV / MediaPipe','A gesture-controlled canvas for building in 3D.'),
    ('web','03','Palestine Now','PRODUCT INTERFACE','React / Vite / Tailwind CSS','An Arabic-first web experience, designed in RTL.'),
    ('ml','04','Bank Marketing ML','APPLIED MACHINE LEARNING','Python / scikit-learn / Jupyter','From preprocessing to classification and evaluation.'),
]


def project(spec,mobile=False):
    kind,number,title,category,stack,desc=spec
    w,h=(600,368) if mobile else (1200,242)
    b = label(30,39,number+' / '+category,LIME,12)
    if mobile:
        b += '<g transform="translate(269 26) scale(.87)">'+illustration(kind)+'</g>'
        b += text(30,202,title,33,PAPER,700,'letter-spacing="-1"')
        b += text(30,241,desc,18,MUTED)
        b += line(30,273,570,273)
        b += text(30,313,stack,16,MUTED)
        b += text(538,317,'↗',28,LIME)
    else:
        b += text(30,101,title,42,PAPER,700,'letter-spacing="-1.5"')
        b += text(30,141,desc,21,MUTED)
        b += text(30,205,stack,16,LIME)
        b += '<g transform="translate(805 28)">'+illustration(kind)+'</g>'
        b += text(1142,205,'↗',30,LIME)
    svg(f'project-{kind}'+('-mobile' if mobile else '')+'.svg',w,h,title,b,desc+' '+stack+'. Decorative illustration; not a product screenshot.')


def stack(mobile=False):
    rows=[('01','LANGUAGES','Python / JavaScript / TypeScript / C++ / SQL'),
          ('02','WEB & BACKEND','React / FastAPI / Django / MySQL / Supabase'),
          ('03','AI & WORKFLOW','OpenCV / MediaPipe / scikit-learn / Git')]
    w,h=(600,360) if mobile else (1200,216)
    b=''
    for i,(n,area,tools) in enumerate(rows):
        y=47+i*(112 if mobile else 66)
        b+=label(28,y,n,LIME,12)+label(75,y,area,MUTED,12)
        b+=text(28 if mobile else 340,y+36 if mobile else y,tools,19 if mobile else 22,PAPER)
        if i<2: b+=line(28,y+(58 if mobile else 24),w-28,y+(58 if mobile else 24))
    svg('stack'+('-mobile' if mobile else '')+'.svg',w,h,'Tools I work with',b,
        'Languages: Python, JavaScript, TypeScript, C++, SQL. Web and backend: React, FastAPI, Django, MySQL, Supabase. AI and workflow: OpenCV, MediaPipe, scikit-learn, Git.')


def contact(mobile=False):
    w,h=(600,245) if mobile else (1200,224)
    b=label(30,43,'HAVE SOMETHING WORTH BUILDING?',LIME,12)
    b+=text(28,112 if mobile else 116,"Let's make it happen.",43 if mobile else 66,PAPER,700,'letter-spacing="-2"')
    b+=text(30,167 if mobile else 179,'Internships / Junior roles / Collaboration',20,MUTED)
    b+=text(530 if mobile else 1080,206 if mobile else 139,'↗',45 if mobile else 82,LIME,400,'class="float"')
    svg('connect'+('-mobile' if mobile else '')+'.svg',w,h,"Let's make it happen — contact Zaid",b)


def main():
    OUT.mkdir(parents=True,exist_ok=True)
    for mobile in (False,True):
        hero(mobile)
        stack(mobile)
        contact(mobile)
        for item in PROJECTS:
            project(item,mobile)
    print(f'Built {len(list(OUT.glob("*.svg")))} SVG assets in {OUT}')


if __name__=='__main__':
    main()

File: scripts/build_artwork.mjs
// Generate the profile artwork with Node.js 18+. No dependencies.
import { mkdirSync, writeFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';

function buildArtwork() {
  const files = {};
  const C = { ink: "#141414", panel: "#1c1c1c", paper: "#f2efe8", muted: "#aaa69e", line: "#363431", orange: "#ff794b" };
  const escape = value => String(value).replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;").replaceAll('"', "&quot;");
  const t = (x,y,value,size=20,color=C.paper,weight=400,extra="") => `<text x="${x}" y="${y}" font-family="Arial,Helvetica,sans-serif" font-size="${size}" font-weight="${weight}" fill="${color}" ${extra}>${escape(value)}</text>`;
  const label = (x,y,value,color=C.muted,size=12) => t(x,y,value,size,color,500,'letter-spacing="1.7"');
  const box = (x,y,w,h,fill=C.panel,r=12,stroke="none") => `<rect x="${x}" y="${y}" width="${w}" height="${h}" rx="${r}" fill="${fill}" stroke="${stroke}"/>`;
  const path = (d,color=C.line,width=1,extra="") => `<path d="${d}" stroke="${color}" stroke-width="${width}" fill="none" ${extra}/>`;
  const dot = (x,y,r,color=C.orange) => `<circle cx="${x}" cy="${y}" r="${r}" fill="${color}"/>`;
  const pill = (x,y,value,width) => box(x,y,width,30,"#242321",15,C.line)+t(x+14,y+20,value,12,C.paper);
  function save(name,w,h,title,body,description=title) {
    files[`assets/profile/${name}.svg`] = `<svg xmlns="http://www.w3.org/2000/svg" width="${w}" height="${h}" viewBox="0 0 ${w} ${h}" role="img" aria-labelledby="title desc">
<title id="title">${escape(title)}</title><desc id="desc">${escape(description)}</desc>
<defs><clipPath id="bounds"><rect width="${w}" height="${h}" rx="18"/></clipPath><pattern id="dots" width="24" height="24" patternUnits="userSpaceOnUse"><circle cx="2" cy="2" r="1" fill="#44413c"/></pattern><linearGradient id="metal" x1="0" y1="0" x2="1" y2="1"><stop stop-color="#fffdf7"/><stop offset=".48" stop-color="#96918b"/><stop offset="1" stop-color="#e9e4da"/></linearGradient></defs>
<style>.float{animation:float 8s ease-in-out infinite}.pulse{animation:pulse 5s ease-in-out infinite}@keyframes float{50%{transform:translateY(-6px)}}@keyframes pulse{50%{opacity:.35}}@media(prefers-reduced-motion:reduce){*{animation:none!important}}</style>
<g clip-path="url(#bounds)">${box(0,0,w,h,C.ink,18)}${body}</g>${box(.5,.5,w-1,h-1,"none",18,C.line)}</svg>\n`;
  }
  // Project a twisted ribbon and sort its faces by depth for a metallic sculpture.
  function sculpture(cx,cy,scale=1) {
    const point = (u,v) => {
      const radius=139+v*Math.cos(u/2), x=radius*Math.cos(u), y=radius*Math.sin(u), z=v*Math.sin(u/2);
      return [x*.87+z*.49, y*.66-z*.66-x*.18, y*.5+z*.75];
    };
    const faces=[];
    const steps=96, strips=10;
    for(let i=0;i<steps;i++) for(let j=0;j<strips;j++) {
      const u=i*Math.PI*2/steps, du=Math.PI*2/steps, v=-48+j*96/strips, dv=96/strips;
      const points=[point(u,v),point(u+du,v),point(u+du,v+dv),point(u,v+dv)];
      const light=Math.max(0,Math.min(1,.5+.35*Math.sin(u*2-.7)+.16*j/strips));
      const rgb=[Math.round(85+168*light),Math.round(82+166*light),Math.round(78+159*light)];
      const color=j===0||j===strips-1?C.orange:`rgb(${rgb.join(",")})`;
      faces.push({depth:points.reduce((s,p)=>s+p[2],0)/4,svg:`<path d="M${points.map(p=>`${p[0].toFixed(2)} ${p[1].toFixed(2)}`).join("L")}Z" fill="${color}" stroke="${color}" stroke-width=".5"/>`});
    }
    faces.sort((a,b)=>a.depth-b.depth);
    return `<g transform="translate(${cx} ${cy}) scale(${scale})"><ellipse cy="162" rx="125" ry="11" fill="#0b0b0b"/><circle r="215" fill="none" stroke="${C.line}"/><circle r="188" fill="none" stroke="${C.line}" stroke-dasharray="2 10"/>${path("M-236 0H-215M215 0H236M0-236V-215M0 215V236","#747068")}${dot(215,0,4)}<g class="float">${faces.map(f=>f.svg).join("")}</g>${label(-50,232,"BUILD / REFINE",C.muted,10)}</g>`;
  }
  function cube(x,y,s=35) {
    return `<g transform="translate(${x} ${y})"><path d="M0 ${-s}L${s} ${-s/2} 0 0 ${-s} ${-s/2}Z" fill="#fff2df"/><path d="M${-s} ${-s/2}L0 0V${s}L${-s} ${s/2}Z" fill="#d04c2b"/><path d="M0 0L${s} ${-s/2}V${s/2}L0 ${s}Z" fill="#ff976c"/></g>`;
  }
  // Decorative miniatures explain each project's domain without inventing product data.
  function illustration(kind) {
    let b=box(0,0,440,248,"#e9e5dc",14);
    if(kind==="api") {
      b+=label(22,28,"REQUEST → DATA", "#66615a",10);
      for(const [x,y,name] of [[28,69,"API"],[171,69,"AUTH"],[314,69,"SQL"]]) {
        b+=box(x,y,98,72,"#f8f5ef",10,"#c6bfb3")+t(x+49,y+43,name,18,C.ink,700,'text-anchor="middle"');
      }
      b+=path("M126 105H171M269 105H314","#8d8375",2);
      b+=path("M77 143V177H363V143","#8d8375",1.5);
      b+=dot(149,177,5,C.orange)+box(103,194,234,28,"#ddd6ca",6)+t(220,213,"MODELS · PERMISSIONS · MIGRATIONS",9,"#534b41",500,'text-anchor="middle"');
    } else if(kind==="voxel") {
      b=box(0,0,440,248,"#b74428",14);
      for(let i=0;i<7;i++) {
        b+=path(`M${62+i*29} ${104+i*14.5}L${222+i*29} ${24+i*14.5}`,"#cf6345");
        b+=path(`M${62+i*29} ${104-i*14.5}L${222+i*29} ${184-i*14.5}`,"#cf6345");
      }
      b+=cube(195,142,36)+cube(267,142,36)+cube(231,124,36);
      b+=`<g class="float">${cube(195,70,36)}</g>`;
      b+=path("M114 67H94V87M302 172H322V152","#ffe6d2",2);
      b+=label(22,226,"HAND TRACKING / 3D CANVAS","#ffe6d2",10);
    } else if(kind==="web") {
      b+=box(44,27,352,193,"#faf8f2",10,"#bdb5a8")+path("M44 54H396","#d8d0c5");
      b+=dot(59,41,3,"#9a9489")+dot(71,41,3,"#c4b9a7")+dot(83,41,3,C.orange);
      b+=box(286,69,91,10,C.ink,3)+box(223,89,154,5,"#a39b8c",2)+box(269,101,108,5,"#bcb4a5",2);
      b+=box(64,120,312,23,"#ebe5d9",5);
      for(let i=0;i<3;i++){
        const x=64+i*108;
        b+=box(x,154,95,48,i===2?"#cf6948":"#d3cabb",5);
        b+=path(`M${x+9} 191L${x+30} 173 ${x+43} 184 ${x+66} 162 ${x+84} 187`,i===2?"#fce7d6":"#a09686",2);
      }
    } else {
      b=box(0,0,440,248,"#262522",14);
      const layers=[[62,[74,124,174]],[168,[52,100,148,196]],[274,[76,124,172]],[377,[100,148]]];
      layers.forEach(([x,ys],i)=>{
        if(i<layers.length-1)for(const y of ys)for(const ny of layers[i+1][1]) b+=path(`M${x} ${y}L${layers[i+1][0]} ${ny}`,"#51483e");
      });
      b+=path("M62 124L168 100 274 124 377 100",C.orange,2);
      layers.forEach(([x,ys],i)=>ys.forEach(y=>{b+=dot(x,y,8,"#262522")+`<circle cx="${x}" cy="${y}" r="7" fill="none" stroke="${i===3?C.orange:"#b8a68b"}" stroke-width="2"/>`;}));
      b+=label(22,27,"PATTERNS → PREDICTIONS",C.muted,10);
      b+=label(22,230,"PREPROCESS / COMPARE / EVALUATE",C.muted,10);
    }
    return b;
  }
  const projects=[
    ["api","01","Books Management API","BACKEND ENGINEERING","FastAPI / MySQL / JWT","Relational data. Deliberate design."],
    ["voxel","02","Pinch Voxel Studio","COMPUTER VISION","Python / OpenCV / MediaPipe","Your hands become the interface."],
    ["web","03","Palestine Now","WEB DEVELOPMENT","React / Vite / Tailwind CSS","Local discovery. An Arabic-first experience."],
    ["ml","04","Bank Marketing ML","APPLIED MACHINE LEARNING","Python / scikit-learn / Jupyter","Finding useful patterns in the data."]
  ];
  for(const mobile of [false,true]) {
    const w=mobile?600:1200,suffix=mobile?"-mobile":"";
    let b=label(36,41,"ZM / ZAID MAYYALLEH",C.paper,12)+label(w-(mobile?132:210),41,mobile?"PALESTINE":"DEVELOPER PORTFOLIO",C.muted,10);
    b+=path(`M36 64H${w-36}`);
    b+=label(40,117,"FULL-STACK / BACKEND / APPLIED AI",C.orange,mobile?11:12);
    b+=t(32,mobile?218:255,"Zaid",mobile?115:151,C.paper,750,'letter-spacing="-7"');
    b+=t(34,mobile?292:351,"Mayyalleh.",mobile?77:96,C.paper,750,'letter-spacing="-5"');
    b+=t(40,mobile?337:408,"Built on curiosity.",mobile?27:32,C.paper,400);
    b+=t(40,mobile?370:446,"Refined through code.",mobile?27:32,C.muted,400);
    b+=sculpture(mobile?300:929,mobile?600:302,mobile?.79:1.06);
    b+=pill(40,mobile?821:505,"BUILD",80)+pill(130,mobile?821:505,"LEARN",83)+pill(223,mobile?821:505,"ITERATE",96);
    b+=path(`M36 ${mobile?878:574}H${w-36}`);
    b+=label(40,mobile?909:607,"AN-NAJAH NATIONAL UNIVERSITY",C.muted,10);
    b+=dot(w-188,mobile?905:603,3,C.orange)+label(w-176,mobile?909:607,"ALWAYS BUILDING",C.paper,10);
    save("hero"+suffix,w,mobile?935:630,"Zaid Mayyalleh — Full-stack developer in Palestine",b,"Built on curiosity. Refined through code. Full-stack development, backend systems, and applied AI. A metallic twisted ribbon accompanies the developer's name; subtle motion respects reduced-motion preferences.");
    for(const [kind,n,title,category,stack,desc] of projects) {
      b=label(32,40,n+" / "+category,C.orange,11);
      if(mobile) {
        b+=`<g transform="translate(32 66) scale(1.21818)">${illustration(kind)}</g>`;
        b+=t(30,417,title,35,C.paper,700,'letter-spacing="-1.1"');
        b+=t(32,453,desc,21,C.muted);
        b+=path("M32 480H568");
        b+=t(32,514,stack,17,C.paper)+t(540,519,"↗",28,C.orange);
      } else {
        b+=t(30,110,title,43,C.paper,700,'letter-spacing="-1.5"');
        b+=t(32,156,desc,22,C.muted);
        b+=t(32,230,stack,17,C.paper);
        b+=label(32,285,"EXPLORE REPOSITORY",C.muted,10)+t(218,288,"↗",19,C.orange);
        b+=`<g transform="translate(726 32)">${illustration(kind)}</g>`;
      }
      save("project-"+kind+suffix,w,mobile?545:314,title+" — "+stack,b,desc+" Decorative project illustration, not a product screenshot.");
    }
    const groups=[
      ["01","Languages",["Python · JavaScript · TypeScript","C++ · SQL"]],
      ["02","Web & backend",["React · FastAPI · Django","MySQL · Supabase"]],
      ["03","AI & workflow",["OpenCV · MediaPipe","scikit-learn · Git"]]
    ];
    b="";
    groups.forEach(([n,title,tools],i)=>{
      const x=mobile?28:28+i*389,y=mobile?28+i*166:28;
      b+=box(x,y,mobile?544:366,145,C.panel,10,C.line);
      b+=label(x+20,y+28,n,C.orange,11);
      b+=t(x+56,y+29,title,20,C.paper,600);
      b+=t(x+20,y+81,tools[0],mobile?22:18,C.muted);
      b+=t(x+20,y+113,tools[1],mobile?22:18,C.muted);
    });
    save("stack"+suffix,w,mobile?526:201,"Tools across projects and ongoing learning",b,"Python, JavaScript, TypeScript, C++, SQL, React, FastAPI, Django, MySQL, Supabase, OpenCV, MediaPipe, scikit-learn, and Git. Tools used and explored, not proficiency ratings.");
    b=box(0,0,w,mobile?329:294,C.paper,18)+label(32,39,"NEXT / LET'S CONNECT","#615b52",11);
    b+=t(28,mobile?112:134,"Have an idea?",mobile?59:86,C.ink,700,'letter-spacing="-3"');
    b+=t(30,mobile?175:215,"Let's build it.",mobile?59:86,C.ink,700,'letter-spacing="-3"');
    b+=t(32,mobile?232:262,"Internships · Junior roles · Collaboration",mobile?19:20,"#615b52");
    b+=box(w-90,mobile?249:101,58,58,C.orange,29)+t(w-61,mobile?290:143,"↗",38,C.ink,400,'text-anchor="middle"');
    save("connect"+suffix,w,mobile?329:294,"Let's connect — contact Zaid",b,"Open to internships, junior roles, and collaboration. Email Zaid Mayyalleh.");
  }
  return files;
}

for (const [path, content] of Object.entries(buildArtwork())) {
  const target = new URL('../' + path, import.meta.url);
  mkdirSync(fileURLToPath(new URL('.', target)), { recursive: true });
  writeFileSync(target, content);
}
